### DISCUSSION TITLE
Hetero GNN Link prediction

### DISCUSSION THREAD

**Discussion by talhaubaid** [pyg-team/pytorch_geometric#4643] [Q&A] [answered]
problem statement: I m new to pyg and I need a recommendation system using a graph neural network. We have to recommend skills to users based on their current performance. For an instant, if a user had already taken some skills and is current if we is taking X skill and their proficiency with X skill is 70 percent (or perform well with the skill x ) then he/she should be shown the next post skill of that X. if he/she does not perform well than recommend the previous skill of X.
I have the information related to post and pre of skill in the adjacency matrix (graph form), skill node feature, and user skill interaction data.
Which GNN algorithm do you suit this kind of problem
![Untitled](https://user-images.githubusercontent.com/96561218/168425126-0b73d3e5-13d3-4d28-815a-248aa5c8969e.png)

**Comment by rusty1s** [comment, marked as answer]
This looks like a heterogeneous link prediction problem to me. You can find an example for this [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/hetero_link_pred.py).

**Reply by talhaubaid** [reply]
do we have any algorithm that will solved the weighted link prediction problem ?

**Reply by talhaubaid** [reply]
I have seen uber eat used the SAGE graph and converted it into the weighted graph for the recommendation system. They just gave the overview and the code is not available for that as well. Is there any algorithm like that in Pytorch (weighted link prediction)?

**Reply by rusty1s** [reply]
The example shown above is actually a weighted link prediction. It uses `mse_loss` to predict the rating of a movie given by a user.

**Reply by talhaubaid** [reply]
@rusty1s Is this example also for a directed graph?

**Reply by rusty1s** [reply]
Yes

**Reply by talhaubaid** [reply]
@rusty1s [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/hetero_link_pred.py) can you have a readme or explanation of this work somewhere? In this what is the point of adding reverse relation for message passing?

**Reply by rusty1s** [reply]
Adding reverse relations lets the message passing phase pass messages in both directions. Otherwise, only destination node types will get updated, while source nodes stay with their original representation. It is a common strategy to boost performance in directional graphs (using distinct parameters for each direction).

**Reply by talhaubaid** [reply]
@rusty1s I m having this error while splitting the train, test, and validation edges
![image](https://user-images.githubusercontent.com/96561218/169042772-1d11c074-5e9e-40b3-8e25-ff16665635ee.png)

**Reply by talhaubaid** [reply]
@rusty1s I have changed the way of data passing to edge_type, re_edge_type parameter. Now its giving me this issue. 
![image](https://user-images.githubusercontent.com/96561218/169051810-0e90a706-a882-4873-a64f-f2c278677177.png)

**Reply by rusty1s** [reply]
I think this is already fixed in master. Can you try to follow [these instructions](https://github.com/pyg-team/pytorch_geometric#nightly-and-master) and test again? Sorry for the inconvenience!

**Reply by talhaubaid** [reply]
@rusty1s thanks it worked with the master. Now I have this issue 
![image](https://user-images.githubusercontent.com/96561218/169300892-8d4391e9-04ab-4301-b593-66d227b8afda.png)
